File: SG/WordHandler.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SG
{
    public static class WordHandler
    {
        private class WordTreeNode
        {
            public char NodeChar;

            public WordTreeNode Parent;
            public List<WordTreeNode> Children;

            public WordTreeNode(char c, WordTreeNode parent = null, List<WordTreeNode> children = null)
            {
                NodeChar = c;
                Parent = parent;
                Children = children != null ? children : new List<WordTreeNode>();
            }

            public bool ShallowContains(char c)
            {
                foreach (WordTreeNode node in Children)
                {
                    if (node.NodeChar == c)
                        return true;
                }

                return false;
            }

            public WordTreeNode ShallowGetNode(char c)
            {
                foreach (WordTreeNode node in Children)
                {
                    if (node.NodeChar == c)
                        return node;
                }
                return null;
            }

            public static bool operator ==(WordTreeNode w1, WordTreeNode w2)
            {
                if (object.ReferenceEquals(null, w1) && object.ReferenceEquals(null, w2))
                    return true;
                if (!object.ReferenceEquals(null, w1) && object.ReferenceEquals(null, w2))
                    return false;
                if (object.ReferenceEquals(null, w1) && !object.ReferenceEquals(null, w2))
                    return false;

                if (w1.NodeChar == w2.NodeChar)
                    return true;
                return false;
            }

            public static bool operator !=(WordTreeNode w1, WordTreeNode w2)
            {
                if (object.ReferenceEquals(null, w1) && object.ReferenceEquals(null, w2))
                    return false;
                if (!object.ReferenceEquals(null, w1) && object.ReferenceEquals(null, w2))
                    return true;
                if (object.ReferenceEquals(null, w1) && !object.ReferenceEquals(null, w2))
                    return true;

                if (w1.NodeChar != w2.NodeChar)
                    return true;
                return false;
            }

            public override bool Equals(object obj)
            {
                if (obj is WordTreeNode && this.NodeChar == ((WordTreeNode)obj).NodeChar)
                    return true;
                return false;
            }
        }

        private class WordTree
        {
            private WordTreeNode Root;

            public WordTree()
            {
                Root = new WordTreeNode('\0');
            }

            public void AddWord(string word)
            {
                WordTreeNode current = Root;
                WordTreeNode next;
                for (int i = 0; i < word.Length; i++)
                {
                    char c = char.ToUpper(word[i]);
                    next = current.ShallowGetNode(c);

                    if (next == null)
                    {
                        next = new WordTreeNode(c, current);
                        current.Children.Add(next);
                    }

                    current = next;
                }

                if (!current.ShallowContains('\0'))
                    current.Children.Add(new WordTreeNode('\0', current));
            }

            public string TypeWord()
            {
                List<WordTreeNode> Pointers = new List<WordTreeNode>();

                while (true)
                {
                    ConsoleKeyInfo key = IOUtils.ReadKey();
                    char c = char.ToUpper(key.KeyChar);

                    if (key.Key == ConsoleKey.Escape)
                        return null;

                    for (int i = 0; i < Pointers.Count; i++)
                    {
                        if (Pointers[i].ShallowContains(c))
                        {
                            Pointers[i] = Pointers[i].ShallowGetNode(c);

                            if (Pointers[i].ShallowContains('\0'))
                                return GetWord(Pointers[i]);
                        }
                        else
                        {
                            Pointers.RemoveAt(i);
                            i--;
                        }
                    }

                    if (Root.ShallowContains(c))
                        Pointers.Add(Root.ShallowGetNode(c));
                }
            }

            private string GetWord(WordTreeNode word)
            {
                if (word == null || word.NodeChar == '\0')
                    return "";
                return GetWord(word.Parent) + word.NodeChar;
            }
        }

        private static Dictionary<string, Word> Words;
        private static WordTree Characters;

        static WordHandler()
        {
            Words = new Dictionary<string, Word>();
            Characters = new WordTree();
        }

        public static void AddWord(Word word)
        {
            Characters.AddWord(word.WordName);

            if (!Words.ContainsKey(word.WordName))
                Words.Add(word.WordName.ToUpper(), word);
        }

        public static Word TypeWord()
        {
            string word = Characters.TypeWord();

            if (word == null)
                return null;

            if (Words.ContainsKey(word))
                return Words[word];

            return null;
        }
    }
}

File: SG/WordHandler.cpp
#include "Word.h"
#include "WordHandler.h"
#include <algorithm>

WordHandler::WordHandler()
{
}

WordHandler::~WordHandler()
{
	auto iter = m_words.begin();

	while (iter != m_words.end())
	{
		delete iter->second;
		iter++;
	}
}

void WordHandler::addWord(Word* word)
{
	std::transform(word->m_word.begin(), word->m_word.end(), word->m_word.begin(), towlower);
	m_characters.addWord(word->m_word);
	m_words[word->m_word] = word;
}

Word* WordHandler::nextChar(wchar_t c)
{
	c = towlower(c);
	std::wstring word = m_characters.nextChar(c);
	if (word.empty())
		return NULL;

	return m_words[word];
}

WordHandler::WordTreeNode::WordTreeNode(wchar_t c, WordTreeNode* parent)
: m_nodechar(c), m_parent(parent)
{
}

WordHandler::WordTreeNode::~WordTreeNode()
{
	auto iter = m_children.begin();
	while (iter != m_children.end())
	{
		if (*iter != nullptr)
		{
			delete (*iter);
			(*iter) = nullptr;
		}
		iter++;
	}
}

bool WordHandler::WordTreeNode::shallowContains(wchar_t c) const
{
	c = towlower(c);
	auto iter = m_children.begin();
	while (iter != m_children.end())
	{
		if ((*iter)->m_nodechar == c)
			return true;
		iter++;
	}
	return false;
}

WordHandler::WordTreeNode* WordHandler::WordTreeNode::shallowGetNode(wchar_t c)
{
	c = towlower(c);
	auto iter = m_children.begin();
	while (iter != m_children.end())
	{
		if ((*iter)->m_nodechar == c)
			return (*iter);
		iter++;
	}
	return NULL;
}

bool WordHandler::WordTreeNode::operator==(const WordTreeNode& node)
{
	if (this->m_nodechar == node.m_nodechar)
		return true;
	return false;
}

bool WordHandler::WordTreeNode::operator!=(const WordTreeNode& node)
{
	if (this->m_nodechar == node.m_nodechar)
		return false;
	return true;
}

WordHandler::WordTree::WordTree()
: m_root(L'\0')
{
}

WordHandler::WordTree::~WordTree()
{
	m_pointers.clear();
}

void WordHandler::WordTree::addWord(std::wstring word)
{
	WordTreeNode* current = &m_root;
	WordTreeNode* next = NULL;

	for (int i = 0; i < word.length(); i++)
	{
		next = current->shallowGetNode(word[i]);
		if (next == NULL)
		{
			next = new WordTreeNode(word[i], current);
			current->m_children.push_back(next);
		}
		current = next;
	}

	if (!current->shallowContains(L'\0'))
		current->m_children.push_back(new WordTreeNode(L'\0', current));
}

std::wstring WordHandler::WordTree::nextChar(wchar_t c)
{
	auto iter = m_pointers.begin();
	while (iter != m_pointers.end())
	{
		if ((*iter)->shallowContains(c))
		{
			(*iter) = (*iter)->shallowGetNode(c);

			if ((*iter)->shallowContains(L'\0'))
			{
				WordTreeNode* word = *iter;
				m_pointers.clear();
				return getWord(word);
			}
		}
		else
		{
			iter = m_pointers.erase(iter);
		}
		iter++;
	}

	if (m_root.shallowContains(c))
		m_pointers.push_back(m_root.shallowGetNode(c));

	return L"";
}

std::wstring WordHandler::WordTree::getWord(WordTreeNode* node)
{
	if (node == NULL || node->m_nodechar == L'\0')
		return L"";
	return getWord(node->m_parent) + node->m_nodechar;
}